### DISCUSSION TITLE
Gulp isn't running tasks in a series

### DISCUSSION THREAD

**Discussion by emily-roses** [gulpjs/gulp#2473] [Help] [answered]
I'm trying to write a function that will delete files in dist that no longer exist in src. When I run the task "clean" below with "gulp clean" on the command line, I only get "clean" as console output, and no files are created. This means that readSrcFiles, readDistFiles, and parseFiles aren't running for some reason. Why?

```
task("clean", function () {
        console.log("clean");
        task("default", series(readSrcFiles, readDistFiles, parseFiles));
        del(["temp"]);
        return Promise.resolve("cleaned");
    });

function readSrcFiles() {
    console.log("readSrc");
    return src("./src/**/*")
        .pipe(filelist("_filelist_src.json", { relative: true }))
        .pipe(dest("./temp"));
}

function readDistFiles() {
    console.log("readDist");
    return src("./name.github.io/**/*")
        .pipe(filelist("_filelist_dist.json", { relative: true }))
        .pipe(dest("./temp"));
}
function parseFiles() {
    console.log("parseFiles");
    let srcFiles = JSON.parse(fs.readFileSync("./temp/_filelist_src.json"));
    let distFiles = JSON.parse(fs.readFileSync("./temp/_filelist_dist.json"));
    distFiles.forEach((element) => {
        if (!srcFiles.includes(element)) {
            console.log(del.sync(["name.github.io/" + element]));
        }
    });
}
```

At first I didn't have this split up into functions, but then it would only run the src command that was being returned, which obviously didn't let me do anything afterwards.

**Comment by phated** [comment, marked as answer]
@EmilyBonar You shouldn't be putting `task()` calls inside another task; there would be no way for gulp to know that a task would be registered inside.

I highly recommend you read through the "Getting started" guide on our website which explains tasks and async completion. It will also show you to use exports instead of the `task` function which isn't recommended anymore because people end up in situations like you've described.

**Comment by emily-roses** [comment]
I had read it, but after some more digging I ended up with this, which works.

```
task(
	"default",
	series(
		"html_imports",
		parallel(readSrcFiles, readDistFiles),
		parseFiles /* , "serve" */
	)
);

function readSrcFiles(res) {
	console.log("src");
	return src("./src/**/*")
		.pipe(filelist("_filelist_src.json", { relative: true }))
		.pipe(dest("./temp"));
}

function readDistFiles(res) {
	console.log("dist");
	return src("./EmilyBonar.github.io/**/*")
		.pipe(filelist("_filelist_dist.json", { relative: true }))
		.pipe(dest("./temp"));
}
function parseFiles(res) {
	console.log("parse");
	let srcFiles = JSON.parse(fs.readFileSync("./temp/_filelist_src.json"));
	let distFiles = JSON.parse(fs.readFileSync("./temp/_filelist_dist.json"));
	distFiles.forEach((element) => {
		if (!srcFiles.includes(element)) {
			console.log(del.sync(["EmilyBonar.github.io/" + element]));
		}
	});
	del.sync(["temp"]);
	return Promise.resolve("cleaned");
}
```
This still leaves me with the smaller problem that readSrcFiles and readDistFiles copy their whole folders instead of just writing the json I want, but at least it'll work for now.

**Reply by phated** [reply]
At this point, you are just using gulp as a globbing library against your files. Why wouldn't you just use a globbing library since they don't read any files? Remember that gulp is just javascript and you can use any libraries that are suitable for the job! There's no need to shoehorn gulp's filesystem access into weird use cases it wasn't made for.

**Comment by emily-roses** [comment]
I'm still pretty new to JavaScript, so if there's a better way I'm all
ears. The main thing I started using Gulp for is to inject
headers/footers/etc into my static site. That meant I had src and dist
folders, so I needed to manage the differences between them. The live
server is nice too.

**Reply by emily-roses** [reply]
I transitioned to using glob for this task, thanks for your suggestions @phated
